sklearn/neighbors/tests/test_neighbors.py

```diff
@@ -1498,6 +1498,66 @@ def test_neighbors_validate_parameters(Estimator):
         nbrs.predict([[]])
 
 
+@pytest.mark.parametrize(
+    "Estimator",
+    [
+        neighbors.KNeighborsClassifier,
+        neighbors.RadiusNeighborsClassifier,
+        neighbors.KNeighborsRegressor,
+        neighbors.RadiusNeighborsRegressor,
+    ],
+)
+@pytest.mark.parametrize("n_features", [2, 100])
+def test_neighbors_minkowski_semimetric(Estimator, n_features):
+    """
+    Validation of all classes extending NeighborsBase with
+    Minkowski semi-metrics (i.e. when 0 < p < 1).
+    """
+    X = rng.random_sample((10, n_features))
+    y = np.ones(10)
+
+    model = Estimator(p=0)
+    with pytest.raises(ValueError):
+        # This Exception will be raised by _validate_params called in fit method
+        model.fit(X, y)
+
+    model = Estimator(p=0.1, algorithm="auto")
+    msg = (
+        "Mind that for 0 < p < 1, Minkowski metrics are not distance metric. Continuing"
+        " the execution with `algo='brute'`."
+    )
+    with pytest.warns(UserWarning, match=msg):
+        model.fit(X, y)
+
+    assert model._fit_method == "brute"
+
+    model = Estimator(p=0.1, algorithm="brute")
+    msg = (
+        "Mind that for 0 < p < 1, Minkowski metrics are not distance metric. Continuing"
+        " the execution with `algo='brute'`."
+    )
+    with pytest.warns(UserWarning, match=msg):
+        model.fit(X, y)
+
+    model = Estimator(algorithm="kd_tree", p=0.1)
+    msg = (
+        'algo="kd_tree" does not support 0 < p < 1 for '
+        "the Minkowski metric. To resolve this problem either "
+        'set p >= 1 or algo="brute".'
+    )
+    with pytest.raises(ValueError, match=msg):
+        model.fit(X, y)
+
+    model = Estimator(algorithm="ball_tree", p=0.1)
+    msg = (
+        'algo="ball_tree" does not support 0 < p < 1 for '
+        "the Minkowski metric. To resolve this problem either "
+        'set p >= 1 or algo="brute".'
+    )
+    with pytest.raises(ValueError, match=msg):
+        model.fit(X, y)
```

Comment:
I think that we should make a separate test. It will allow to parameterized it and to have less redundant code.

Reply:
@glemaitre can you elaborate more on this?🤔

Reply:
Something like that:

```python
@pytest.mark.parametrize(
    "Estimator",
    [
        neighbors.KNeighborsClassifier,
        neighbors.RadiusNeighborsClassifier,
        neighbors.KNeighborsRegressor,
        neighbors.RadiusNeighborsRegressor,
    ],
)
@pytest.mark.parametrize("algorithm", ["kd_tree", "ball_tree"])
def test_neighbors_minkowski_algo_error(Estimator, algorithm):
    """Check that we raise a proper error if `algorithm!='brute'` and `p<1`."""
    X = rng.random_sample((10, 2))
    y = np.ones(10)

    model = Estimator(algorithm=algorithm, p=0.1)
    msg = (
        f'algorithm="{algorithm}" does not support 0 < p < 1 for '
        "the Minkowski metric. To resolve this problem either "
        'set p >= 1 or algorithm="brute".'
    )
    with pytest.raises(ValueError, match=msg):
        model.fit(X, y)
```

Reply:
And the idea is to do something similar for code that is repeated in the test.